pkg/api/file.go

```diff
@@ -0,0 +1,265 @@
+// Copyright 2020 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package api
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"mime"
+	"mime/multipart"
+	"net/http"
+	"strconv"
+
+	"github.com/ethersphere/bee/pkg/collection/entry"
+	"github.com/ethersphere/bee/pkg/file"
+	"github.com/ethersphere/bee/pkg/file/joiner"
+	"github.com/ethersphere/bee/pkg/jsonhttp"
+	"github.com/ethersphere/bee/pkg/storage"
+	"github.com/ethersphere/bee/pkg/swarm"
+	"github.com/gorilla/mux"
+)
+
+const (
+	MultiPartFormData = "multipart/form-data"
+)
+
+type FileUploadResponse struct {
+	Reference swarm.Address `json:"reference"`
+}
+
+// bzzFileUploadHandler uploads the file and its metadata supplied as a multipart http message.
+func (s *server) bzzFileUploadHandler(w http.ResponseWriter, r *http.Request) {
+	contentType, params, err := mime.ParseMediaType(r.Header.Get("Content-Type"))
+	if contentType != MultiPartFormData {
+		s.Logger.Debugf("file: no mutlipart: %v", err)
+		s.Logger.Error("file: no mutlipart")
+		jsonhttp.BadRequest(w, "not a mutlipart/form-data")
+		return
+	}
+
+	mr := multipart.NewReader(r.Body, params["boundary"])
+	for {
+		part, err := mr.NextPart()
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			s.Logger.Debugf("file: read mutlipart: %v", err)
+			s.Logger.Error("file: read mutlipart")
+			jsonhttp.BadRequest(w, "error reading a mutlipart/form-data")
+			return
+		}
+
+		ctx := r.Context()
+
+		// try to find filename
+		// 1) in part header params
+		// 2) as formname
+		// 3) file reference hash (after uploading the file)
+		fileName := part.FileName()
+		if fileName == "" {
+			fileName = part.FormName()
+		}
+
+		// then find out content type
+		contentType := part.Header.Get("Content-Type")
+		reader := bufio.NewReader(part)
+		if contentType == "" {
+			buf, err := reader.Peek(512)
+			if err != nil && err != io.EOF {
+				s.Logger.Debugf("file: read content type: %v, file name %s", err, fileName)
+				s.Logger.Error("file: read content type")
+				jsonhttp.BadRequest(w, "error reading content type")
+				return
+			}
+			contentType = http.DetectContentType(buf)
+		}
+
+		// find file size
+		fileSizeString := part.Header.Get("Content-Length")
+		if fileSizeString == "" {
+			s.Logger.Debugf("file: content length: %v", err)
```

Comment:
This error is nil here. The log message should be about missing content length.

Currently, I get no content length for a part when I submit with the form like this one:

```
<form action="http://localhost:8080/file" method="post" enctype="multipart/form-data">
 <div>
   <label for="file">Choose file to upload</label>
   <input type="file" id="file" name="file">
 </div>
 <div>
   <button>Submit</button>
 </div>
</form>
```

Do you maybe have an idea what is wrong? As I had this form working with the code from last week, which is not here anymore because of the force push.

Reply:
I have created a pr that handles this like we do in swarm already https://github.com/ethersphere/bee/pull/296. With it, the upload is working nicely.

Reply:
merged your PR